src/providers/coinbase.rs

```diff
@@ -0,0 +1,137 @@
+    tracing::{info, log::error},
+    url::Url,
+};
+
+#[derive(Debug)]
+pub struct CoinbaseProvider {
+    pub api_key: String,
+    pub app_id: String,
+    pub http_client: Client<HttpsConnector<hyper::client::HttpConnector>>,
+}
+
+impl CoinbaseProvider {
+    pub fn new(api_key: String, app_id: String) -> Self {
+        let http_client = Client::builder().build::<_, hyper::Body>(HttpsConnector::new());
+        Self {
+            api_key,
+            app_id,
+            http_client,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct CoinbaseResponseBody {
+    pub transactions: Vec<CoinbaseTransaction>,
+    pub next_page_key: Option<String>,
+    pub total_count: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone)]
+pub struct CoinbaseTransaction {
+    pub status: String,
+    pub transaction_id: String,
+    pub tx_hash: String,
+    pub created_at: String,
+}
+
+#[async_trait]
+impl HistoryProvider for CoinbaseProvider {
+    async fn get_transactions(
+        &self,
+        address: H160,
+        body: Bytes,
+        params: HistoryQueryParams,
+    ) -> RpcResult<HistoryResponseBody> {
+        let base = format!(
+            "https://pay.coinbase.com/api/v1/buy/user/{}/transactions",
+            &address.to_string()
+        );
+        let mut url = Url::parse(&base).map_err(|_| RpcError::HistoryParseCursorError)?;
+        url.query_pairs_mut().append_pair("page_size", "50");
+
+        if let Some(cursor) = params.cursor {
+            url.query_pairs_mut().append_pair("page_key", &cursor);
+        }
+
+        error!(
+            "Coinbase URL: {:?}, {:?}",
+            url.as_str(),
+            self.app_id.clone()
+        );
+
+        let hyper_request = hyper::http::Request::builder()
+            .uri(url.as_str())
+            .method(method::Method::GET)
+            .header("Content-Type", "application/json")
+            .header("CBPAY-APP-ID", self.app_id.clone())
+            .header("CBPAY-API-KEY", self.api_key.clone())
+            .body(hyper::body::Body::from(body))?;
+
+        let response = self.http_client.request(hyper_request).await?;
+
+        if !response.status().is_success() {
+            error!(
+                "Error on Coinbase transactions response. Status is not OK: {:?}",
+                response.status(),
+            );
+            return Err(RpcError::TransactionProviderError);
+        }
+
+        let mut body = response.into_body();
+        let mut bytes = Vec::new();
+        while let Some(next) = body.next().await {
+            bytes.extend_from_slice(&next?);
+        }
+        let body: CoinbaseResponseBody = match serde_json::from_slice(&bytes) {
+            Ok(body) => body,
+            Err(e) => {
+                error!("Error on parsing coinbase transactions response: {:?}", e);
+                return Err(RpcError::TransactionProviderError);
+            }
+        };
```

Comment:
`.map_err()` and `.tap_err()` and turbofish is nicer IMO

```suggestion
        let body = serde_json::from_slice::<CoinbaseResponseBody>(&bytes)
            .tap_err(|e| error!("Error on parsing coinbase transactions response: {e:?}"))
            .map_err(|_| RpcError::TransactionProviderError)?;
```